### PR TITLE
Build with CUDA 12.6

### PR DESCRIPTION
**Pull request by jeongseok-meta**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by jeongseok-meta** [comment]
@conda-forge-admin, please rerender

**Comment by jeongseok-meta** [comment]
@hmaarrfk I lost from the first step. It seems we can only use CUDA versions listed [here](https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/main/recipe/conda_build_config.yaml#L56-L59), but it doesn't have 12.2.

**Comment by jeongseok-meta** [comment]
@conda-forge-admin, please rerender

**Comment by jeongseok-meta** [comment]
@conda-forge-admin, please rerender

File: recipe/conda_build_config.yaml
Comment (hmaarrfk) [inline recipe/conda_build_config.yaml:42]: you'll want to effectively copy this section for `linux64`

**Comment by jeongseok-meta** [comment]
@conda-forge-admin, please rerender

**Comment by jeongseok-meta** [comment]
@conda-forge-admin, please rerender

File: recipe/conda_build_config.yaml
Comment (hmaarrfk) [inline recipe/conda_build_config.yaml:48]:
```suggestion
  - 12.2                                            # [linux64]
```

Comment (hmaarrfk) [reply recipe/conda_build_config.yaml:48]: you might need to replicate the rest of everything too.

**Review by hmaarrfk — APPROVED** [review]

File: recipe/conda_build_config.yaml
Comment (jeongseok-meta) [reply recipe/conda_build_config.yaml:48]: Rerendering with 12.2 seems to fail per:

> If there is a recipe/conda-build-config.yaml file in the feedstock make sure that it is compatible with the current [global pinnnings](https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/master/recipe/conda_build_config.yaml).

**Comment by jeongseok-meta** [comment]
@conda-forge-admin, please rerender

**Comment by hmaarrfk** [comment]
👏🏾

**Comment by hmaarrfk** [comment]
>     "error": "User not authorized for the requested runners, user not present in 'users' list."

set it to azure for now, and see how far you get ;)

**Comment by hmaarrfk** [comment]
You might also want to rebase onto https://github.com/conda-forge/pytorch-cpu-feedstock/pull/261/commits/22bf05548d61de9a30bff2f60c83ff2990168485

File: conda-forge.yml
Comment (hmaarrfk) [inline conda-forge.yml:23]:
```suggestion
  linux_64: azure
```

**Comment by jeongseok-meta** [comment]
@conda-forge-admin, please rerender

**Comment by jeongseok-meta** [comment]
@conda-forge-admin, please restart ci

File: conda-forge.yml
Comment (hmaarrfk) [inline conda-forge.yml:4]:
```suggestion
```

**Comment by hmaarrfk** [comment]
@conda-forge-admin please rerender

**Comment by hmaarrfk** [comment]
the "yaml" files control the CI files, so you always have to rerender after changing them.

**Comment by hmaarrfk** [comment]
i'm speachless...

File: conda-forge.yml
Comment (hmaarrfk) [inline conda-forge.yml:2]:
```suggestion
```

Maybe try to remove this for now.

Comment (hmaarrfk) [reply conda-forge.yml:2]: (and of course, rerender)

**Comment by jeongseok-meta** [comment]
@conda-forge-admin please rerender

File: .scripts/run_docker_build.sh
Comment (hmaarrfk) [inline .scripts/run_docker_build.sh:1]: Its really weird that this file just got deleted. but yet a .ci_support file does exist.

Comment (hmaarrfk) [reply .scripts/run_docker_build.sh:1]: maybe just checkout this file from the other branch?

Comment (jeongseok-meta) [reply .scripts/run_docker_build.sh:1]: It seems a glitch of the diff view since the file still exists: https://github.com/jeongseok-meta/pytorch-cpu-feedstock/blob/b6df3b5/.scripts/run_docker_build.sh

Comment (jeongseok-meta) [reply .scripts/run_docker_build.sh:1]: and still see the same error (of permission):

```
/home/vsts/work/_temp/8b47f76b-a122-455d-aafa-e1924c3fbbad.sh: line 12: .scripts/run_docker_build.sh: Permission denied
```

Comment (jeongseok-meta) [reply .scripts/run_docker_build.sh:1]: Locally:

```
➜  pytorch-cpu-feedstock git:(cuda122) ll .scripts                    
total 24
-rw-r--r--  1 jeongseok  staff   3.5K Oct  4 14:16 build_steps.sh
-rw-r--r--  1 jeongseok  staff   887B Sep 29 08:17 logging_utils.sh
-rw-r--r--  1 jeongseok  staff   3.8K Oct  4 14:16 run_docker_build.sh
```

Comment (jeongseok-meta) [reply .scripts/run_docker_build.sh:1]: Interesting, it says it's an empty file (when it's not) if the change is only a file permission change.

image

**Comment by jeongseok-meta** [comment]
Now it's building! However, I'm still wondering how to enforce the use of CUDA 12.2 when setting `cuda_compiler_version` to 12.2 fails during rerendering due to:

> If there is a recipe/conda-build-config.yaml file in the feedstock make sure that it is compatible with the current [global pinnnings](https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/master/recipe/conda_build_config.yaml).

**Comment by hmaarrfk** [comment]
ok so my tip is to "delete the whole outputs section" from the meta.yaml

then on a powerful machine, use

```
CPU_COUNT=10 python build-locally.py --debug
```
to get into an interactive shell and find the missing dependencies.

**Comment by jeongseok-meta** [comment]
@conda-forge-admin, please rerender

**Comment by hmaarrfk** [comment]
feel free to squash all my junk commits. i would leave the rerender commits alone, or drop them when you rebase!

**Comment by jeongseok-meta** [comment]
@conda-forge-admin, please rerender

**Comment by jeongseok-meta** [comment]
It seems that the recipe [doesn't work for finding CUDA 12.0](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=1055507&view=logs&j=7b6f2c87-f3a7-5133-8d84-7c03a75d9dfc&t=9eb77fd2-8ddd-5444-8fc0-71cb28dcb736&l=1571), even with the patches:

```
  CMake Error at cmake/public/cuda.cmake:70 (message):
    Failed to find nvToolsExt
  Call Stack (most recent call first):
    cmake/Dependencies.cmake:43 (include)
    CMakeLists.txt:857 (include)
```

Now that [PyTorch 2.5.0 has been released](https://github.com/pytorch/pytorch/releases/tag/v2.5.0), which includes upstream patches for CUDA 12, I will wait for a version update of this recipe in hopes that it resolves the CUDA finding issue for CUDA 12.

**Comment by hmaarrfk** [comment]
> Now that [PyTorch 2.5.0 has been released](https://github.com/pytorch/pytorch/releases/tag/v2.5.0),

feel free to make the PR for the update ;)

**Comment by jeongseok-meta** [comment]
@conda-forge-admin, please rerender

**Comment by Tobias-Fischer** [comment]
For `nvToolsExt` on Windows, you might want to take a look at https://github.com/Tobias-Fischer/kaldi-feedstock/blob/91a882031d77296d9d47efb751d8cf43632df7c7/recipe/meta.yaml#L95-L96 and https://github.com/Tobias-Fischer/kaldi-feedstock/blob/91a882031d77296d9d47efb751d8cf43632df7c7/recipe/patches/0006-Cuda-12-support.patch#L113-L121

**Comment by Tobias-Fischer** [comment]
But this might be relevant:
```
2024-10-17T20:57:32.4709616Z   -- Could NOT find nvtx3 (missing: nvtx3_dir)
2024-10-17T20:57:32.4710074Z   CMake Warning at cmake/public/cuda.cmake:180 (message):
2024-10-17T20:57:32.4710508Z     Cannot find NVTX3, find old NVTX instead
2024-10-17T20:57:32.4711225Z   Call Stack (most recent call first):
2024-10-17T20:57:32.4711671Z     cmake/Dependencies.cmake:44 (include)
2024-10-17T20:57:32.4712106Z     CMakeLists.txt:863 (include)
```
(from https://github.com/pytorch/pytorch/blob/7a117f3b3eea4cfeef21da2e3a8a1e39c30fa07d/cmake/public/cuda.cmake#L172-L183)

**Comment by Tobias-Fischer** [comment]
I think we've missing https://github.com/conda-forge/nvtx-c-feedstock

**Comment by jeongseok-meta** [comment]
@conda-forge-admin, please rerender

**Comment by hmaarrfk** [comment]
you will only get 360 minutes. no matter what.

As a final test, you can try to compile locally.

If you have a workstation (which i think you must ;) ) then it should take about 1-2 hours with docker

```
python build-locally.py
```

**Comment by hmaarrfk** [comment]
However, one question might be whether or not it is worth bumping all the way to 12.6
https://github.com/conda-forge/cupy-feedstock/blob/main/.ci_support/migrations/cuda12x.yaml#L70

File: recipe/meta.yaml
Comment (hmaarrfk) [inline recipe/meta.yaml:196]: it might be time to undo all these changes and clean up?

**Comment by jeongseok-meta** [comment]
> However, one question might be whether or not it is worth bumping all the way to 12.6
> https://github.com/conda-forge/cupy-feedstock/blob/main/.ci_support/migrations/cuda12x.yaml#L70

Yeah, I'm down for bumping all the way to 12.6. 12.2 was the minimal required version for our cases, but bumping up to the latest should be fine (let me know if this is not the case though).

**Comment by jeongseok-meta** [comment]
> As a final test, you can try to compile locally.
> 
> If you have a workstation (which i think you must ;) ) then it should take about 1-2 hours with docker

This sounds good. Let me try. However, how would we proceed when it builds locally (maybe some of the Linux variants but not tested them all) but we don't know whether CI passes or not due to the timeout (and even after merging, how can CI build and upload to the repository)? Sorry, I'm still learning the conda-forge workflow :)

**Comment by hmaarrfk** [comment]
> how would we proceed when it builds locally

I can make a small "dummy commit" to your branch and trigger things.

Or eventually, given your tennacity (this is really the requirement), you can become a maintainer here and some more powers get unlocked especially with these runners.

However, they are scarce even for us, so I would have gone through the same process of moving to azure, then building locally.

**Comment by jeongseok-meta** [comment]
@conda-forge-admin, please rerender

**Comment by jeongseok-meta** [comment]
> Or eventually, given your tennacity (this is really the requirement), you can become a maintainer here and some more powers get unlocked especially with these runners.

I apologize for missing this earlier, but I'd be honored to become a maintainer here and unlock more powers with the runners!

**Comment by jslee02** [comment]
linux_64 generic and mkl with CUDA 12.6: [linux_64_blas_impl.zip](https://github.com/user-attachments/files/17601188/linux_64_blas_impl.zip)

linux_aarch64 is still building...

**Comment by jslee02** [comment]
Just a heads up, I'm building the Linux packages locally of this PR, similar to what I did for https://github.com/conda-forge/pytorch-cpu-feedstock/pull/277

**Comment by hmaarrfk** [comment]
do you want to rebase onto main and get the commit history in sync?

I would prefer if you were able to restart all builds after you rebase + bump the build numbers.

File: recipe/conda_build_config.yaml
Comment (hmaarrfk) [reply recipe/conda_build_config.yaml:42]: do we want to build aarch64 with 12.6 as well?

**Comment by hmaarrfk** [comment]
So few questions:

1. Do you want to build 12.6 for arm as well?
2. Do you want to try ppc64le with GPU 12.6?

**Comment by jeongseok-meta** [comment]
At the moment, I only need CUDA 12.6 for Linux 64-bit. In the long term, I would also like to support other versions, but there is no specific timeline or concrete use case yet. That being said I'd prefer to add only CUDA 12.6 for Linux 64-bit.

**Comment by jeongseok-meta** [comment]
> do you want to rebase onto main and get the commit history in sync?
> 
> I would prefer if you were able to restart all builds after you rebase + bump the build numbers.

Done! Building now... ⏳

**Comment by hmaarrfk** [comment]
PS. given such little differences between the this and build 0, i would accept only builds with cuda 12.6

**Comment by jeongseok-meta** [comment]
> given such little differences between the this and build 0, i would accept only builds with cuda 12.6

Sounds good to me. Let me then build only for cuda 12.6 to move fast.

**Comment by jslee02** [comment]
Build log: [log_files.zip](https://github.com/user-attachments/files/17612105/log_files.zip)

Packages (`*-cuda126*.conda`):
- https://anaconda.org/jslee02/libtorch
- https://anaconda.org/jslee02/pytorch
- https://anaconda.org/jslee02/pytorch-cpu
- https://anaconda.org/jslee02/pytorch-gpu

**Comment by hmaarrfk** [comment]
> Packages (*-cuda126*.conda):

I actually cannot use this selector when uploading. 'Labels' work well. could you add the `cfep03` to the new packages please?

**Comment by jeongseok-meta** [comment]
Do you want the label for cuda126 or including all the other pytorch 2.5.1?

**Comment by hmaarrfk** [comment]
just the new packages.

**Comment by hmaarrfk** [comment]
i think if you label all i can use the `--skip-existing` flag to not recopy the files over

**Comment by jeongseok-meta** [comment]
Done!

**Comment by hmaarrfk** [comment]
Sorry shoudln't there be a aarch cuda 12.6 package as well?

```
anaconda copy --from-label cfep03 --to-label main --to-owner conda-forge jslee02/libtorch/2.5.1
anaconda copy --from-label cfep03 --to-label main --to-owner conda-forge jslee02/pytorch/2.5.1
anaconda copy --from-label cfep03 --to-label main --to-owner conda-forge jslee02/pytorch-cpu/2.5.1
anaconda copy --from-label cfep03 --to-label main --to-owner conda-forge jslee02/pytorch-gpu/2.5.1
```

**Comment by jeongseok-meta** [comment]
No, I intended to add only linux64 (not aarch) to move faster. But if I knew we could update this quickly, I would add them here too. How would you like to proceed? We could add them in this PR or in a follow-up. Either way works for me.

**Comment by hmaarrfk** [comment]
lets do it here without any force push. I want the OSX builds to finish building on main.

**Comment by hmaarrfk** [comment]
The cuda 12.6 builds have been uploaded already for linux 64. give them a test drive... you know just in case.

**Comment by hmaarrfk** [comment]
you can probably delete everything from your channel too for space.....

**Comment by jeongseok-meta** [comment]
> lets do it here without any force push. I want the OSX builds to finish building on main.

Let me build locally for aarch64 + cuda126 without push a commit, as I expect the OSX build will be retriggered anyway, and then push the change for aarch64 after the current OSX build is done.

**Comment by jeongseok-meta** [comment]
- Deleted all the packages from my channel
- Initiated aarch64 + cuda126 build locally
- Will push the changes once the current OSX builds are done

**Comment by hmaarrfk** [comment]
you really don't need to worry about this branch, it was mostly about the 'main' branch where the packages get uploaded.

**Comment by jslee02** [comment]
Building aarch64 + cuda126 is done!

Build log: [log_files.zip](https://github.com/user-attachments/files/17612574/log_files.zip)

Packages (only new ones are uploaded):
- https://anaconda.org/jslee02/libtorch
- https://anaconda.org/jslee02/pytorch
- https://anaconda.org/jslee02/pytorch-gpu

**Comment by hmaarrfk** [comment]
I'm not sure if this is important but things I observd:

```
  -- Unable to find cublas_v2.h in either "/home/conda/feedstock_root/build_artifacts/libtorch_1730670943607/_build_env/targets/sbsa-linux/include" or "/home/conda/feedstock_root/build_artifacts/libtorch_1730670943607/_build_env/math_libs/include"
```

**Comment by hmaarrfk** [comment]
```
anaconda copy --from-label main --to-label main --to-owner conda-forge jslee02/libtorch/2.5.1
anaconda copy --from-label main --to-label main --to-owner conda-forge jslee02/pytorch/2.5.1
anaconda copy --from-label main --to-label main --to-owner conda-forge jslee02/pytorch-gpu/2.5.1
```

**Comment by hmaarrfk** [comment]
Please feel free to issue the command:
https://conda-forge.org/docs/maintainer/infrastructure/#conda-forge-admin-please-add-user-username

with the appropriate username !

**Comment by jeongseok-meta** [comment]
@conda-forge-admin, please add user @jeongseok-meta

**Comment by hmaarrfk** [comment]
you might need to do this on an issue.

**Pull request merged**
